cql3/untyped_result_set.cc

```diff
@@ -47,58 +47,118 @@
-       return std::make_pair<sstring, bytes_opt>(c->name->to_string(), std::move(d));
-    });
-    return tmp;
-}())
-{}
+size_t cql3::untyped_result_set_row::index(const std::string_view& name) const {
+    auto i = _name_to_index->find(name);
+    return i != _name_to_index->end() ? i->second : std::numeric_limits<size_t>::max();
+}
 
 bool cql3::untyped_result_set_row::has(std::string_view name) const {
-    auto i = _data.find(name);
-    return i != _data.end() && i->second;
+    auto i = index(name);
+    if (i < _data.size()) {
+        return !std::holds_alternative<std::monostate>(_data.at(i));
+    }
+    return false;
+}
+
+cql3::untyped_result_set_row::view_type cql3::untyped_result_set_row::get_view(std::string_view name) const {
+    return std::visit(make_visitor(
+        [](std::monostate) -> view_type { throw std::bad_variant_access(); },
+        [](const view_type& v) -> view_type { return v; },
+        [](const bytes& b) -> view_type { return view_type(b); }
+    ), _data.at(index(name)));
+}
+
+const std::vector<lw_shared_ptr<cql3::column_specification>>& cql3::untyped_result_set_row::get_columns() const {
+    return _metadata.get_names();
 }
 
 using cql_transport::messages::result_message;
 
-cql3::untyped_result_set::untyped_result_set(const cql3::result_set& rs) {
-    auto& cn = rs.get_metadata().get_names();
-    for (auto& r : rs.rows()) {
-        // r is const ref. TODO: make this more efficient by either wrapping result set
-        // or adding modifying accessors to it.
-        _rows.emplace_back(cn, r);
-    }
+cql3::untyped_result_set_row::index_map_ptr cql3::untyped_result_set::make_index(const cql3::metadata& metadata) {
+    auto map = ::make_shared<untyped_result_set_row::index_map>();
+    auto& names = metadata.get_names();
+    size_t i = 0;
+    std::transform(names.begin(), names.end(), std::inserter(*map, map->end()), [&](lw_shared_ptr<column_specification> c) mutable {
+        return std::make_pair<std::string_view, size_t>(c->name->text(), i++);
+    });
+    return map;
 }
 
+struct cql3::untyped_result_set::visitor {
+    rows_type& rows;
+    const cql3::metadata& meta;
+    untyped_result_set_row::index_map_ptr index;
+    untyped_result_set_row::data_views tmp;
+
+    visitor(rows_type& r, const cql3::metadata& m, untyped_result_set_row::index_map_ptr i)
+        : rows(r)
+        , meta(m)
+        , index(i)
+    {}
+
+    void start_row() {
+        rows.reserve(index->size());
+    }
+    void accept_value(std::optional<query::result_bytes_view>&& v) {
+        if (v) {
+            tmp.emplace_back(std::move(*v));
+        } else {
+            tmp.emplace_back(std::monostate{});
+        } 
+    }
+    // somewhat weird dispatch, but when visiting directly via
+    // result_generator, pk:s will be temporary - and sent 
+    // as views, not opt_views. So we can dispatch on this and 
+    // simply copy the temporaries.
```

Comment:
Would it be possible to move from those temporaries instead of copying from them?

Reply:
Maybe. Not sure if that violates the contract.

Reply:
No, you cannot, since the args are views, not the actual temporaries. And changing that in turn would break other code, which I guess is fixable, but probably not worth it... At least not for now.